### ISSUE TITLE
array_index_out_of_bounds_exception for field name which is "."

### ISSUE THREAD

**Issue by robin13** [elastic/elasticsearch#21862]
**Elasticsearch version**: 5.0.1

**Plugins installed**: [] none

**JVM version**:1.8.0_111

**OS version**: Ubuntu 16.04

**Description of the problem including expected versus actual behavior**:

**Steps to reproduce**:
Without any custom templates or mapping, try this:

```
POST /test/test/1
{
    "." : {
        "foo" : {
            "bar" : 123
        }
    }
}
```

and you will receive this 400 error:
```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "mapper_parsing_exception",
        "reason" : "failed to parse"
      }
    ],
    "type" : "mapper_parsing_exception",
    "reason" : "failed to parse",
    "caused_by" : {
      "type" : "array_index_out_of_bounds_exception",
      "reason" : "-1"
    }
  },
  "status" : 400
}
```

It seems that a single `.` is not a valid field name (I cannot find documentation detailing this), and even if this is the case, the error is not very helpful.

**Comment by jpountz** [comment]
This seems to be a bug in the way we handle bots in field names.

**Comment by clintongormley** [comment]
> This seems to be a bug in the way we handle bots in field names.
We allow bots in field names??!?

**Comment by jpountz** [comment]
Woops!

**Comment by MaineC** [comment]
This looks like it might be related to a discussion over on discuss from last week:

https://discuss.elastic.co/t/name-cannot-be-empty-string-error-if-field-name-starts-with/66808/2

**Comment by hariso** [comment]
Something which is interesting for me, `foo.` works as a field name. Expected?

**Comment by danlevy1** [comment]
I know this bug has been open for a while, but I am looking more into it again. I was looking at the discussion here: [https://discuss.elastic.co/t/name-cannot-be-empty-string-error-if-field-name-starts-with/66808/2](https://discuss.elastic.co/t/name-cannot-be-empty-string-error-if-field-name-starts-with/66808/2) based on @MaineC 's comment. I am getting conflicting information: some are saying it's a bug and others are saying it's just a documentation problem. Can some Elasticsearch developers advise me where they are on this issue?

**Comment by clintongormley** [comment]
@danlevy1 It's not a bug that a field may not be called `.`. It is a bug that we throw an AOOB exception - we should return a more meaningful exception.

**Comment by danlevy1** [comment]
It looks like I have been able to fix this bug! The error I now throw is: 

```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "mapper_parsing_exception",
        "reason" : "failed to parse"
      }
    ],
    "type" : "mapper_parsing_exception",
    "reason" : "failed to parse",
    "caused_by" : {
      "type" : "illegal_argument_exception",
      "reason" : "object field containing only a [.] makes the object resolution ambiguous: ['.']"
    }
  },
  "status" : 400
}
```

Is there anything that you want me to change about the error reason @clintongormley?

**Comment by javanna** [comment]
@elastic/es-search-aggs

**Comment by jtibshirani** [comment]
As a note, we may want to consider #28948 as well when thinking through a fix.

**Comment by javanna** [comment]
This has been addressed in 7.13, see #70972.
